File: nodejs/src/handlers/events.ts
export interface ArtifactsFile {
  filename: string | null;
  size: number | null;
}

export interface Author {
  name: string;
  email: string;
}

export interface Build {
  id: number;
  stage: string;
  name: string;
  status: string;
  created_at: string;
  started_at?: string;
  finished_at?: string;
  when: string;
  manual: boolean;
  user: User;
  runner: unknown;
  artifacts_file: ArtifactsFile;
}

export interface Changes {
  updated_by_id: {
    previous: number | null;
    current: number;
  };
  updated_at: {
    previous: string;
    current: string;
  };
  labels: {
    previous: Label[];
    current: Label[];
  };
}

export interface Commit {
  id: string;
  message: string;
  title: string;
  timestamp: string;
  url: string;
  author: Author;
}

export interface Issue {
  id: number;
  title: string;
  assignee_ids: number[];
  assignee_id: number;
  author_id: number;
  project_id: number;
  created_at: string;
  updated_at: string;
  position: number;
  branch_name: string | null;
  description: string;
  milestone_id: number | null;
  state: string;
  iid: number;
  labels: Label[];
}

export interface IssueAttributes {
  id: number;
  title: string;
  assignee_ids: number[];
  assignee_id: number;
  author_id: number;
  project_id: number;
  created_at: string;
  updated_at: string;
  updated_by_id: number;
  last_edited_at: string | null;
  last_edited_by_id: number | null;
  relative_position: number;
  description: string;
  milestone_id: number;
  state_id: number;
  confidential: boolean;
  discussion_locked: boolean;
  due_date: string | null;
  moved_to_id: number | null;
  duplicated_to_id: number | null;
  time_estimate: number;
  total_time_spent: number;
  time_change: number;
  human_total_time_spent: number | null;
  human_time_estimate: number | null;
  human_time_change: number | null;
  weight: number | null;
  iid: number;
  url: string;
  state: string;
  action: "open" | "close" | "reopen" | "update";
  severity: string;
  escalation_status: string;
  escalation_policy: {
    id: number;
    name: string;
  };
  labels: Label[];
}

export interface Label {
  id: number;
  title: string;
  color: string;
  project_id: number;
  created_at: string;
  updated_at: string;
  template: boolean;
  description: string;
  type: string;
  group_id: number;
}

export interface LastCommit {
  id: string;
  message: string;
  title: string;
  timestamp: string;
  url: string;
  author: Author;
}

export interface MergeRequest {
  id: number;
  target_branch: string;
  source_branch: string;
  source_project_id: number;
  author_id: number;
  assignee_id: number;
  title: string;
  created_at: string;
  updated_at: string;
  milestone_id: number;
  state: string;
  merge_status: string;
  target_project_id: number;
  iid: number;
  description: string;
  position: number;
  labels: Label[];
  url: string;
  source: Project;
  target: Project;
  last_commit: LastCommit;
  work_in_progress: boolean;
  assignee: User;
  detailed_merge_status: string;
}

export interface MergeRequestAttributes {
  assignee_id: number;
  author_id: number;
  created_at: string;
  description: string;
  head_pipeline_id: number;
  id: number;
  iid: number;
  last_edited_at: string | null;
  last_edited_by_id: number | null;
  merge_commit_sha: string | null;
  merge_error: unknown;
  merge_params: {
    force_remove_source_branch: "0" | "1";
  };
  merge_status: string;
  merge_user_id: number | null;
  merge_when_pipeline_succeeds: boolean;
  milestone_id: number | null;
  source_branch: string;
  source_project_id: number;
  state_id: number;
  target_branch: string;
  target_project_id: number;
  time_estimate: number;
  title: string;
  updated_at: string;
  updated_by_id: number;
  url: string;
  source: Project;
  target: Project;
  last_commit: LastCommit;
  work_in_progress: boolean;
  total_time_spent: number;
  time_change: number;
  human_total_time_spent: number | null;
  human_time_change: number | null;
  human_time_estimate: number | null;
  assignee_ids: number[];
  labels: Label[];
  state: string | null;
  blocking_discussions_resolved: boolean;
  action: string | null;
}

export interface NoteAttributes {
  id: number;
  note: string;
  noteable_type: string;
  author_id: number;
  created_at: string;
  updated_at: string;
  project_id: number;
  attachment: unknown;
  line_code: string | null;
  commit_id: string;
  noteable_id: number;
  system: boolean;
  st_diff: StDiff | null;
  url: string;
}

export interface PipelineAttributes {
  id: number;
  iid: number;
  ref: string;
  tag: boolean;
  sha: string;
  before_sha: string;
  source: string;
  status: string;
  stages: string[];
  created_at: string;
  finished_at: string;
  duration: number;
  variables: unknown[];
}

export interface Project {
  id: number;
  name: string;
  description: string;
  web_url: string;
  avatar_url: string | null;
  git_ssh_url: string;
  git_http_url: string;
  namespace: string;
  visibility_level: number;
  path_with_namespace: string;
  default_branch: string;
  ci_config_path: string | null;
  homepage: string;
  url: string;
  ssh_url: string;
  http_url: string;
}

export interface Repository {
  name: string;
  url: string;
  description: string;
  homepage: string;
}

export interface Snippet {
  id: number;
  title: string;
  content: string;
  author_id: number;
  project_id: number;
  created_at: string;
  updated_at: string;
  file_name: string;
  expires_at: string | null;
  type: string;
  visibility_level: number;
}

export interface StDiff {
  diff: string;
  new_path: string;
  old_path: string;
  a_mode: string;
  b_mode: string;
  new_file: boolean;
  renamed_file: boolean;
  deleted_file: boolean;
}

export interface User {
  id: number;
  name: string;
  username: string;
  avatar_url: string;
  email: string;
}

export interface Wiki {
  web_url: string;
  git_ssh_url: string;
  git_http_url: string;
  path_with_namespace: string;
  default_branch: string;
}

export interface WikiPageAttributes {
  title: string;
  content: string;
  format: string;
  message: string;
  slug: string;
  url: string;
  action: string;
}

// --------------------------------------------------
// Events
// --------------------------------------------------

export interface BuildEvent {
  object_kind: "build";
  ref: string;
  tag: boolean;
  before_sha: string;
  sha: string;
  build_id: number;
  build_name: string;
  build_stage: string;
  build_status: string;
  build_started_at: string;
  build_finished_at: string;
  build_duration: unknown;
  build_allow_failure: boolean;
  retries_count: number;
  pipelin_id: number;
  project_id: number;
  project_name: string;
  user: User;
  commit: {
    id: string;
    sha: string;
    message: string;
    author_name: string;
    author_email: string;
    status: string;
    duration: unknown;
    started_at: string | null;
    finished_at: string | null;
  };
  repository: Repository;
  runner: unknown;
  environment: unknown;
}

export interface IssueEvent {
  object_kind: "issue";
  event_type: "issue";
  user: User;
  project: Project;
  object_attributes: IssueAttributes;
  repository: Repository;
  assignees: User[];
  assignee: User;
  labels: Label[];
  changes: Changes;
}

export interface MergeRequestEvent {
  object_kind: "merge_request";
  event_type: "merge_request";
  user: User;
  project: Project;
  object_attributes: MergeRequestAttributes;
  labels: Label[];
  changes: unknown;
  repository: Repository;
  assignees: User[];
}

export interface NoteEvent {
  object_kind: "note";
  event_type: "note";
  user: User;
  project_id: number;
  project: Project;
  repository: Repository;
  object_attributes: NoteAttributes;
  commit?: Commit;
  merge_request?: MergeRequest;
  issue?: Issue;
  snippet?: Snippet;
}

export interface PipelineEvent {
  object_kind: "pipeline";
  object_attributes: PipelineAttributes;
  merge_request: unknown;
  user: User;
  project: Project;
  commit: Commit;
  source_pipeline: unknown;
  builds: Build[];
}

export interface PushEvent {
  object_kind: "push";
  event_name: "push";
  before: string;
  after: string;
  ref: string;
  checkout_sha: string;
  user_id: number;
  user_name: string;
  user_username: string;
  user_email: string;
  user_avatar: string;
  project_id: number;
  project: Project;
  repository: Repository;
  commits: Commit[];
  total_commits_count: number;
}

export interface TagPushEvent {
  object_kind: "tag_push";
  event_name: "tag_push";
  before: string;
  after: string;
  ref: string;
  checkout_sha: string;
  user_id: number;
  user_name: string;
  user_avatar: string;
  project_id: number;
  project: Project;
  repository: Repository;
  commits: Commit[];
  total_commits_count: number;
}

export interface WikiPageEvent {
  object_kind: "wiki_page";
  user: User;
  project: Project;
  wiki: Wiki;
  object_attributes: WikiPageAttributes;
}

export type WebhookEvent =
  | BuildEvent
  | IssueEvent
  | MergeRequestEvent
  | NoteEvent
  | PipelineEvent
  | PushEvent
  | TagPushEvent
  | WikiPageEvent;

File: rust/src/events.rs
use std::error::Error;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use serde::{Deserialize, Serialize};

// pub struct ArtifactsFile {
//     filename: Option<String>,
//     size: Option<u32>,
// }

#[derive(Deserialize, Debug, PartialEq)]
pub struct Author {
    name: String,
    email: String,
}

// pub struct Build {
//     id: u32,
//     stage: String,
//     name: String,
//     status: String,
//     created_at: String,
//     started_at: Option<String>,
//     finished_at: Option<String>,
//     when: String,
//     manual: bool,
//     user: User,
//     artifacts_file: ArtifactsFile,
// }
//
//
// pub struct Commit {
//     id: String,
//     message: String,
//     title: String,
//     timestamp: String,
//     url: String,
//     author: Author,
// }
//
// pub struct Issue {
//     id: u32,
//     title: String,
//     assignee_ids: Vec<u32>,
//     assignee_id: u32,
//     author_id: u32,
//     project_id: u32,
//     created_at: String,
//     updated_at: String,
//     position: u32,
//     branch_name: Option<String>,
//     description: String,
//     milestone_id: Option<u32>,
//     state: String,
//     iid: u32,
//     labels: Vec<Label>,
// }
//
// pub struct IssueAttributes {
//     id: u32,
//     title: String,
//     assignee_ids: Vec<u32>,
//     assignee_id: u32,
//     author_id: u32,
//     project_id: u32,
//     created_at: String,
//     updated_at: String,
//     updated_by_id: u32,
//     last_edited_at: Option<String>,
//     last_edited_by_id: Option<u32>,
//     relative_position: u32,
//     description: String,
//     milestone_id: u32,
//     state_id: u32,
//     confidential: bool,
//     discussion_locked: bool,
//     due_date: Option<String>,
//     moved_to_id: Option<u32>,
//     duplicated_to_id: Option<u32>,
//     time_estimate: u32,
//     total_time_spent: u32,
//     time_change: u32,
//     human_total_time_spent: Option<u32>,
//     human_time_estimate: Option<u32>,
//     human_time_change: Option<u32>,
//     weight: Option<u32>,
//     iid: u32,
//     url: String,
//     state: String,
//     action: String,
//     severity: String,
//     escalation_status: String,
//     labels: Vec<Label>,
// }

#[derive(Deserialize, Debug, PartialEq)]
pub struct Label {
    id: u32,
    title: String,
    color: String,
    project_id: u32,
    created_at: String,
    updated_at: String,
    template: bool,
    description: Option<String>,
    r#type: String,
    group_id: Option<u32>,
}

#[derive(Deserialize, Debug, PartialEq)]
pub struct LastCommit {
    id: String,
    message: String,
    timestamp: String,
    url: String,
    author: Author,
}

#[derive(Deserialize, Debug, PartialEq)]
pub struct MergeRequest {
    id: u32,
    target_branch: String,
    source_branch: String,
    source_project_id: u32,
    author_id: u32,
    assignee_id: u32,
    title: String,
    state: String,
    merge_status: String,
    target_project_id: u32,
    iid: u32,
    description: String,
    labels: Vec<Label>,
    url: String,
    source: Project,
    target: Project,
    work_in_progress: bool,
}

// pub struct PipelineAttributes {
//     id: u32,
//     iid: u32,
//     r#ref: String,
//     tag: bool,
//     sha: String,
//     before_sha: String,
//     source: String,
//     status: String,
//     stages: Vec<String>,
//     created_at: String,
//     finished_at: String,
//     duration: u32,
// }

#[derive(Deserialize, Debug, PartialEq)]
pub struct Project {
    id: u32,
    name: String,
    description: String,
    web_url: String,
    avatar_url: Option<String>,
    git_ssh_url: String,
    git_http_url: String,
    namespace: String,
    visibility_level: u32,
    path_with_namespace: String,
    default_branch: String,
    ci_config_path: Option<String>,
    homepage: String,
    url: String,
    ssh_url: String,
    http_url: String,
}

#[derive(Deserialize, Debug, PartialEq)]
pub struct Repository {
    name: String,
    url: String,
    description: String,
    homepage: String,
}

// pub struct Snippet {
//     id: u32,
//     title: String,
//     content: String,
//     author_id: u32,
//     project_id: u32,
//     created_at: String,
//     updated_at: String,
//     file_name: String,
//     expires_at: Option<String>,
//     r#type: String,
//     visibility_level: u32,
// }
//
// pub struct StDiff {
//     diff: String,
//     new_path: String,
//     old_path: String,
//     a_mode: String,
//     b_mode: String,
//     new_file: bool,
//     renamed_file: bool,
//     deleted_file: bool,
// }

#[derive(Deserialize, Debug, PartialEq)]
pub struct User {
    id: u32,
    name: String,
    username: String,
    avatar_url: String,
    email: Option<String>,
}

// pub struct Wiki {
//     web_url: String,
//     git_ssh_url: String,
//     git_http_url: String,
//     path_with_namespace: String,
//     default_branch: String,
// }
//
// pub struct WikiPageAttributes {
//     title: String,
//     content: String,
//     format: String,
//     message: String,
//     slug: String,
//     url: String,
//     action: String,
// }

// --------------------------------------------------
// Events
// --------------------------------------------------

// pub struct BuildEvent {
//     object_kind: String,
//     r#ref: String,
//     tag: bool,
//     before_sha: String,
//     sha: String,
//     build_id: u32,
//     build_name: String,
//     build_stage: String,
//     build_status: String,
//     build_started_at: String,
//     build_finished_at: String,
//     build_allow_failure: bool,
//     retries_count: u32,
//     pipelin_id: u32,
//     project_id: u32,
//     project_name: String,
//     user: User,
//     repository: Repository,
// }
//
// pub struct IssueEvent {
//     object_kind: String,
//     event_type: String,
//     user: User,
//     project: Project,
//     object_attributes: IssueAttributes,
//     repository: Repository,
//     assignees: Vec<User>,
//     assignee: User,
//     labels: Vec<Label>,
//     // changes: Changes,
// }

#[derive(Deserialize, Debug, PartialEq)]
pub struct MergeRequestAttributes {
    assignee_id: Option<u32>,
    author_id: u32,
    description: String,
    iid: u32,
    target_branch: String,
    source_branch: String,
    source_project_id: u32,
    assignee_ids: Vec<u32>,
    title: String,
    state: String,
    blocking_discussions_resolved: bool,
    work_in_progress: bool,
    merge_status: String,
    target_project_id: u32,
    url: String,
    source: Project,
    target: Project,
    labels: Vec<Label>,
    action: Option<String>,
}

#[derive(Deserialize, Debug, PartialEq)]
pub struct MergeRequestEvent {
    event_type: String,
    user: User,
    project: Project,
    repository: Repository,
    object_attributes: MergeRequestAttributes,
    labels: Vec<Label>,
    assignees: Vec<User>,
    reviewers: Vec<User>,
}

#[derive(Deserialize, Debug, PartialEq)]
pub struct NoteAttributes {
    author_id: u32,
    commit_id: Option<String>,
    id: u32,
    note: String,
    noteable_type: String,
    project_id: u32,
    description: String,
    url: String,
}

#[derive(Deserialize, Debug, PartialEq)]
pub struct NoteEvent {
    event_type: String,
    user: User,
    project_id: u32,
    project: Project,
    repository: Repository,
    object_attributes: NoteAttributes,
    // commit: Option<Commit>,
    merge_request: Option<MergeRequest>,
    // issue: Option<Issue>,
    // snippet: Option<Snippet>,
}

// pub struct PipelineEvent {
//     object_kind: String,
//     object_attributes: PipelineAttributes,
//     user: User,
//     project: Project,
//     commit: Commit,
//     builds: Vec<Build>,
// }
//
// pub struct PushEvent {
//     object_kind: String,
//     event_name: String,
//     before: String,
//     after: String,
//     r#ref: String,
//     checkout_sha: String,
//     user_id: u32,
//     user_name: String,
//     user_username: String,
//     user_email: String,
//     user_avatar: String,
//     project_id: u32,
//     project: Project,
//     repository: Repository,
//     commits: Vec<Commit>,
//     total_commits_count: u32,
// }
//
// pub struct TagPushEvent {
//     object_kind: String,
//     event_name: String,
//     before: String,
//     after: String,
//     r#ref: String,
//     checkout_sha: String,
//     user_id: u32,
//     user_name: String,
//     user_avatar: String,
//     project_id: u32,
//     project: Project,
//     repository: Repository,
//     commits: Vec<Commit>,
//     total_commits_count: u32,
// }
//
// pub struct WikiPageEvent {
//     object_kind: String,
//     user: User,
//     project: Project,
//     wiki: Wiki,
//     object_attributes: WikiPageAttributes,
// }
//
// #[derive(Deserialize, Debug)]
// pub enum WebhookEvent {
//     BuildEvent,
//     IssueEvent,
//     MergeRequestEvent,
//     NoteEvent,
//     PipelineEvent,
//     PushEvent,
//     TagPushEvent,
//     WikiPageEvent,
// }
//
// #[derive(Deserialize, Debug)]
// pub struct WebhookEventPayload {
//     body: WebhookEvent,
// }
//
// pub fn event_from_reader<P: AsRef<Path>>(path: P) -> Result<WebhookEventPayload, Box<dyn Error>> {
//     let file = File::open(path)?;
//     let reader = BufReader::new(file);
//     let payload = serde_json::from_reader(reader)?;
//     Ok(payload)
// }
//
// pub fn event_from_str(s: &str) -> Result<WebhookEventPayload, Box<dyn Error>> {
//     let payload = serde_json::from_str(s)?;
//     Ok(payload)
// }

// https://serde.rs/enum-representations.html

#[derive(Deserialize, Debug, PartialEq)]
#[serde(tag = "object_kind")]
pub enum Body {
    #[serde(rename = "note")]
    Note(NoteEvent),
    #[serde(rename = "merge_request")]
    MR(MergeRequestEvent),
}

#[derive(Deserialize, Debug, PartialEq)]
pub struct Payload {
    body: Body,
}

// --------------------------------------------------
// Functions
// --------------------------------------------------

pub fn payload_from_str(s: &str) -> Result<Payload, Box<dyn Error>> {
    let payload = serde_json::from_str(s)?;
    Ok(payload)
}

pub fn payload_from_file<P: AsRef<Path>>(path: P) -> Result<Payload, Box<dyn Error>> {
    let file = File::open(path)?;
    let reader = BufReader::new(file);
    let payload = serde_json::from_reader(reader)?;
    Ok(payload)
}

// --------------------------------------------------
// Tests
// --------------------------------------------------

#[test]
fn it_note_event_from_file() {
    let payload = payload_from_file("../events/note-mr-mergeable.json").unwrap();
    match payload.body {
        Body::Note(_) => assert!(true),
        _ => assert!(false, "not handled")
    }
}

#[test]
fn it_mr_event_from_file() {
    let payload = payload_from_file("../events/mr-open.json").unwrap();
    match payload.body {
        Body::MR(_) => assert!(true),
        _ => assert!(false, "not handled")
    }
}
